exchange/bitswap/wantmanager.go

```diff
@@ -137,36 +137,49 @@ func (mq *msgQueue) runQueue(ctx context.Context) {
 	for {
 		select {
 		case <-mq.work: // there is work to be done
-
-			err := mq.network.ConnectTo(ctx, mq.p)
-			if err != nil {
-				log.Noticef("cant connect to peer %s: %s", mq.p, err)
-				// TODO: cant connect, what now?
-				continue
-			}
-
-			// grab outgoing message
-			mq.outlk.Lock()
-			wlm := mq.out
-			if wlm == nil || wlm.Empty() {
-				mq.outlk.Unlock()
-				continue
-			}
-			mq.out = nil
-			mq.outlk.Unlock()
-
-			// send wantlist updates
-			err = mq.network.SendMessage(ctx, mq.p, wlm)
-			if err != nil {
-				log.Noticef("bitswap send error: %s", err)
-				// TODO: what do we do if this fails?
-			}
+			mq.doWork(ctx)
 		case <-mq.done:
 			return
 		}
 	}
 }
 
+func (mq *msgQueue) doWork(ctx context.Context) {
+	// allow a minute for connections
+	// this includes looking them up in the dht
+	// dialing them, and handshaking
+	conctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
+
+	err := mq.network.ConnectTo(conctx, mq.p)
+	if err != nil {
+		log.Noticef("cant connect to peer %s: %s", mq.p, err)
+		// TODO: cant connect, what now?
+		return
+	}
+
+	// grab outgoing message
+	mq.outlk.Lock()
+	wlm := mq.out
+	mq.out = nil
+	mq.outlk.Unlock()
+
+	if wlm == nil || wlm.Empty() {
+		return
+	}
```

Comment:
hrm, fair. i can put in:
```
if !wlm.Empty() {
    mq.out = nil
}
```
instead of blindly setting it nil. 

Setting it nil here doesnt change the behaviour, not nil'ing it out was just an allocation optimization.